fix(upgrade): make ngUpgrade work with testability API

sjelin opened the pull request:
- **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

(Docs not really relevant in this case)
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This change makes the testability API work with hybrid apps.  Specifically, it fixes deferred bootstrapping and waiting on ng2 components within ng1 apps
- **What is the current behavior?** (You can also link to an open issue here)

Currently deferred bootstrapping causes ngUpgrade to hang, and if an ng2 component uses `setTimeout` or the like the hybrid app's testability checks will totally ignore the fact that its ng2 components aren't yet stable
- **What is the new behavior (if this is a feature change)?**

Now deferred bootstrapping works, and the testability API will check both ng1 and ng2 stability
- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

Only if someone is relying on their hybrid app to crash!
- **Other information**:

I have almost zero typescript knowledge so please tell me how to fix any style problems.  I did compile/run tests though.

sjelin committed:
fix(upgrade): make ngUpgrade work with testability API

sjelin commented:
I signed it!

sjelin commented:
I've signed it before but you can have me resign any time you want @googlebot :wink:

sjelin commented:
cc @juliemr & whoever owns ngUpgrade

juliemr commented on modules/angular2/src/facade/lang.ts:
```diff
@@ -30,6 +30,7 @@ export interface BrowserNodeGlobal {
   clearTimeout: Function;
   setInterval: Function;
   clearInterval: Function;
+  angular?: any;
```
We should probably add a comment that this will only be available if it's using ngUpgrade.

sjelin replied:
Done

juliemr commented on modules/angular2/src/upgrade/upgrade_adapter.ts:
```diff
@@ -331,6 +334,27 @@ export class UpgradeAdapter {
+                var newWhenStable = (callback: Function): void => {
+                  var whenStableContext: any = this;
+                  origonalWhenStable.call(this, function() {
+                    if (ng2Testability.isStable()) {
```
I don't think this is necessary - can we remove the if check and just always do:

`ng2Testability.whenStable(newWhenStable.bind(whenStableContext, callback));`?

sjelin replied:
No that would be an infinite loop

juliemr replied:
Maybe I'm being dense, but how? ng2 whenStable doesn't call ng1 whenStable.

sjelin replied:
ng `whenStable` is calling `newWhenStable`, which in turn calls ng1 `whenStable`.

The easiest way to see this is by looking at the function as you say I should write it:

```js
var newWhenStable = (callback: Function): void => {
  var whenStableContext: any = this;
  origonalWhenStable.call(this, function() {
    ng2Testability.whenStable(newWhenStable.bind(whenStableContext, callback));
  });
});
```

Note that once `newWhenStable` is called, all code paths lead to `newWhenStable` being called again, and nothing actually invokes `callback`.

juliemr replied:
Gotcha, I see it now. Thanks!

gkalpak commented on modules/angular2/src/upgrade/angular_js.ts:
```diff
@@ -97,16 +97,33 @@ export interface IControllerService {
+      module: noNg,
+      element: noNg,
+      version: noNg,
+      resumeBootstrap: noNg,
```
Why not add `resumeBooystrap` in the type above ?

sjelin replied:
It's there.  clang just put it on the name line as `version`

gkalpak replied:
Wow ! Did clang-format really do that ? 
That pretty unexpected !

<sub>(But then again you can't argue with clang-format.)</sub>

sjelin replied:
I mean, *I* didn't put it on that line, so I think it must have been clang

gkalpak commented on modules/angular2/src/upgrade/upgrade_adapter.ts:
```diff
@@ -345,9 +369,28 @@ export class UpgradeAdapter {
+    Promise.all([
+             this.compileNg2Components(compiler, hostViewFactoryRefMap),
+             ng1compilePromise,
+             ng1BootstrapPromise
```
I think `ng1compilePromise` should run after `ng1BootstrapPromise`, or else the extra modules passed via `resumeBootstrap` won't affect the downgraded components.

sjelin replied:
Ok, how should I do that?  Move the run block down below the bootstrapping code?

gkalpak replied:
Without really looking into the code, I would di something like:

```js
ng1BootstrapPromise.
  then(/* ...create and return ng1compilePromise... */)
```

sjelin replied:
I'm getting a lot of errors when I try to bootstrap before doing the compilation stuff

sjelin replied:
I think this is a problem for a future PR.  I've documented the problem here: https://github.com/angular/angular/issues/7635

gkalpak replied:
What did you try ? From a quick look I can't think of a reason why something like `ng1BootstrapPromise.then(() => ng1compilePromise)` would cause issues.

<sub>But if it does, I guess it's OK leaving it for a future PR.</sub>

sjelin replied:
Comment addressed.  Sorry I misunderstood!

sjelin commented:
@mhevery ping

mhevery commented on modules/angular2/src/facade/lang.ts:
```diff
@@ -30,6 +30,7 @@ export interface BrowserNodeGlobal {
   clearTimeout: Function;
   setInterval: Function;
   clearInterval: Function;
+  angular?: any;  // available for apps using ngUpgrade
```
Could we keep angular1 in ngUpgrade only.

sjelin replied:
I could probably do some casting to access the `angular` object without expanding the interface

mhevery commented:
LGTM, except for the small nit-pick of keeping this to `ngUpgraed` only

sjelin commented:
Having a lot of trouble rebasing this, hopefully I'll get it working later today

sjelin commented:
Ok, rebased.  It's breaking travis because of https://github.com/angular/zone.js/issues/283, which I've submitted a PR for and hopefully will be fixed on npm soon.  For now I've changed the implementation of `zone.js` in my `node_modules` and the unit tests do pass, at least locally.

mhevery commented:
Rebasing on top of https://github.com/angular/angular/pull/7700 should make this Green.

sjelin commented:
Rebasing fixed tests.  All comments addressed.

mary-poppins commented:
Merging PR #7603 on behalf of @kara to branch [presubmit-kara-pr-7603](https://github.com/angular/angular/tree/presubmit-kara-pr-7603).
